File: 2021_random_set_self_practice/using_python/신고결과받기.py
from collections import defaultdict


class User(object):
    def __init__(self, name, reporting=None):
        if reporting is None:
            reporting = set()
        self.name = name
        self.reporting = reporting

    def __repr__(self):
        return f"name-{self.name}"


class Pool(object):
    def __init__(self, users, threshold, id_list, reported=None):
        if reported is None:
            reported = defaultdict(set)
        self.id_list = id_list
        self.users = users
        self.threshold = threshold
        self.reported = reported
        self.punished = set()

    def mk_punished_users(self):
        for name, users_reporting in self.reported.items():
            if len(self.reported[name]) >= self.threshold:
                self.punished.add(name)

    def mk_mail(self):
        ret = []
        for id in self.id_list:
            cnt = 0
            for bad in self.punished:
                if bad in self.users[id].reporting:
                    cnt += 1
            ret.append(cnt)

        return ret


def solution(id_list, report, k):
    users = dict()
    for id in id_list:
        users[id] = User(id)

    pools = Pool(users=users, threshold=k, id_list=id_list)
    for info in report:
        a, b = info.split()
        users[a].reporting.add(b)
        pools.reported[b].add(a)
    pools.mk_punished_users()
    return pools.mk_mail()


if __name__ == "__main__":
    print(solution(["muzi", "frodo", "apeach", "neo"],
                   ["muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"], 2))

File: 2021_random_set_self_practice/using_java/SrcCodes/src/신고결과받기/신고결과받기.java
package 신고결과받기;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

class User {
    String name;
    Set<String> reporting;

    public User(String name, Set<String> reporting) {
        this.name = name;
        this.reporting = reporting;
    }
}

class Pool {
    String[] id_list;
    HashMap<String, User> users;
    HashMap<String, Integer> reported_cnt;
    int k;

    public Pool(String[] id_list, HashMap<String, User> users, int k) {
        this.id_list = id_list;
        this.users = users;
        this.k = k;
        this.reported_cnt = new HashMap<>();
        for (String id: id_list){
            this.reported_cnt.put(id, 0);
        }
    }

    public int[] mk_emails() {
        // 신고 당한 횟수 count 올리기
        for (String name: this.users.keySet()) {
            for (String reported: this.users.get(name).reporting) {
                int val = this.reported_cnt.get(reported);
                this.reported_cnt.put(reported, val + 1);
            }
        }

        // 신고 당한 횟수 k 이상인 애들 세어주기
        int[] ret = new int[this.id_list.length];
        int idx = 0;
        for (String name: this.id_list) {
            int cnt = 0;
            for (String reported: this.users.get(name).reporting) {
                if (this.reported_cnt.get(reported) >= this.k) {
                    cnt++;
                }
            }
            ret[idx++] = cnt;
        }
        return ret;
    }
}


class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        HashMap<String, User> users = new HashMap<>();
        for (String id: id_list) {
            users.put(id, new User(id, new HashSet<>()));
        }

        for (String info: report) {
            String[] tmp = info.split(" ");
            String reporting = tmp[0];
            String reported = tmp[1];
            users.get(reporting).reporting.add(reported);
        }
        Pool pool = new Pool(id_list, users, k);
        return pool.mk_emails();
    }

    public static void main(String[] args) {
        String[] id_list = {"muzi", "frodo", "apeach", "neo"};
        String[] report = {"muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi"};
        int k = 2;

        Solution sol = new Solution();
        System.out.println(Arrays.toString(sol.solution(id_list, report, k)));
    }
}